presto-hive/src/main/java/io/prestosql/plugin/hive/CachedDirectoryLister.java

```diff
@@ -0,0 +1,147 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.prestosql.plugin.hive;
+
+import com.google.common.cache.Cache;
+import com.google.common.cache.CacheBuilder;
+import com.google.common.cache.Weigher;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableSet;
+import io.airlift.units.Duration;
+import io.prestosql.plugin.hive.metastore.Table;
+import io.prestosql.spi.connector.SchemaTableName;
+import org.apache.hadoop.fs.FileSystem;
+import org.apache.hadoop.fs.LocatedFileStatus;
+import org.apache.hadoop.fs.Path;
+import org.apache.hadoop.fs.RemoteIterator;
+import org.weakref.jmx.Managed;
+
+import javax.inject.Inject;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Set;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Collectors;
+
+public class CachedDirectoryLister
+        implements DirectoryLister
+{
+    private final Cache<Path, List<LocatedFileStatus>> locatedFileStatusCache;
+    private final Set<SchemaTableName> tableNames;
+
+    @Inject
+    public CachedDirectoryLister(HiveClientConfig hiveClientConfig)
+    {
+        this(hiveClientConfig.getFileStatusCacheExpireAfterWrite(), hiveClientConfig.getFileStatusCacheMaxSize(), hiveClientConfig.getFileStatusCacheTables());
+    }
+
+    private CachedDirectoryLister(Duration expireAfterWrite, long maxSize, List<String> tables)
+    {
+        this.locatedFileStatusCache = CacheBuilder.newBuilder().maximumWeight(maxSize).weigher((Weigher<Path, List<LocatedFileStatus>>) (key, value) -> value.size()).expireAfterWrite(expireAfterWrite.toMillis(), TimeUnit.MILLISECONDS).recordStats().build();
+        this.tableNames = tables == null ? ImmutableSet.of() : tables.stream().map(CachedDirectoryLister::parseTableName).collect(Collectors.toSet());
+    }
+
+    private static SchemaTableName parseTableName(String tableName)
+    {
+        String[] parts = tableName.split("\\.");
+        if (parts.length != 2) {
+            throw new IllegalArgumentException("Invalid schemaTableName " + tableName);
+        }
+        return new SchemaTableName(parts[0], parts[1]);
+    }
+
+    @Override
+    public RemoteIterator<LocatedFileStatus> list(FileSystem fs, Table table, Path path)
+            throws IOException
+    {
+        List<LocatedFileStatus> files = locatedFileStatusCache.getIfPresent(path);
+        if (files != null) {
```

Comment:
When will the cached entry be evicted/invalidated?

If there's a change to the path (e.g., a new file added to a partition) and that path always hit cache in `list()` operation, it seems to me that Presto will never be able to see the newly added files.

Reply:
This cache works best when your ingestion pattern is immutable, e,g, insert new partition to a table. If, insert overwrite a table, the cache could lead to query failure if timeout is longer than ingestion frequency. That's why I added a list of table to cache in config.

I was thinking reader error could feedback cache here to evict stale entries. However, I didn't think of good solution.